runtimes/alibi-explain/mlserver_alibi_explain/explainers/sklearn_api_runtime.py

```diff
@@ -0,0 +1,36 @@
+from typing import Any
+
+import joblib
+import lightgbm as lgb
+from xgboost.core import XGBoostError
+from lightgbm.basic import LightGBMError
+
+from mlserver_xgboost.xgboost import _load_sklearn_interface as load_xgb_model
+from mlserver.errors import InvalidModelURI
+from mlserver_alibi_explain.explainers.white_box_runtime import AlibiExplainWhiteBoxRuntime
+
+
+class AlibiExplainSKLearnAPIRuntime(AlibiExplainWhiteBoxRuntime):
+    """
+    Runtime for white-box explainers that require access to a tree-based model matching the SKLearn API, such as
+    a sklearn, XGBoost, or LightGBM model. Example explainers include TreeShap and TreePartialDependence.
+    """
+    async def _get_inference_model(self) -> Any:
+        inference_model_path = self.alibi_explain_settings.infer_uri
+        # Attempt to load model in order: XGBoost, LightGBM, sklearn
+        # TODO - add support for CatBoost (would require model_type = 'classifier' or 'regressor' in settings)
+        try:
+            # Try to load as sklearn model first
+            model = joblib.load(inference_model_path)
+        except (IndexError, KeyError, IOError):
+            try:
+                # Try to load as XGBoost model
+                model = load_xgb_model(inference_model_path)
+            except XGBoostError:
+                try:
+                    # Try to load as LightGBM model (do this last as raises warning if not successful)
+                    model = lgb.Booster(model_file=inference_model_path)
```

Comment:
Thanks, looking again I think you're right about this. The above would result in a raw `Booster` not an sklearn api wrapped model. 

I initially went with `lgb.Booster(...)` because that is what `mlserver_lightgbm` uses:

https://github.com/SeldonIO/MLServer/blob/3aee441aa8f0d22732c97ee879b46a3a7cbd086c/runtimes/lightgbm/mlserver_lightgbm/lightgbm.py#L22

On the other hand, `mlserver_xgboost` results in sklearn API models being loaded:

https://github.com/SeldonIO/MLServer/blob/3aee441aa8f0d22732c97ee879b46a3a7cbd086c/runtimes/xgboost/mlserver_xgboost/xgboost.py#L47

This difference is tricky for the white-box explainers, since we want the latter (i.e. sklearn models). I see a number of options:
1. See if we can just expect all of these white-box models to be given as `joblib` files (sklearn deffo works, as does `lightgbm`. Would need to test `xgboost`.). This results in a pretty problematic difference to the model runtime's themselves.
2. Use `mlserver_xgboost` etc (related to the discussion of assuming `model-settings.json` exists etc). However, would need to modify `mlserver_lightgbm` to wrap models in the SKLearn API (potentially problematic, requires access to `_` attributes.
3. Just stick to sklearn and xgboost models for now. Leave `lightgbm` until later...

Reply:
Thanks for that context @ascillitoe!

As far as I can see, the ideal solution would be to support both methods in the LightGBM runtime (we'd need both because since the format is different, it would be a breaking change otherwise). We could then use some defaults (and / or settings flags) to decide which method to use in the LGBM runtime (e.g. if file is `model.bst` - use LGBM's `load()`; if file is `model.joblib` - use `joblib`).

The White Box Explainer runtime can then look into that to decide whether it can explain a LGBM model or not (e.g. it could look for a `LGBMRuntime._sklearn_interface: bool` flag that we could add to the LGBM runtime), and raise an error if the LGBM model was saved as a booster?

Reply:
I agree that adding support for `.bst` vs `joblib` to `mlserver_lightgbm` might be nice. Not least because I think there might be a potentially confusing difference between `mlserver_xgboost` and `mlserver_lightgbm` right now. If a user  serialised sklearn api compatible classifier's from each library (as `'xgb_model.json'` and `lightgbm_model.bst`), the `mlserver_xgb` runtime would support `infer_output='predict_proba`, whilst the `mlserver_lightgbm` runtime would not (since the loaded `lgb.Booster` does not). 

Above is unrelated to explainers! Wrt to explainers, I think the main point now is deciding if we want to enforce this `model-setting.json` being present in the model dir, or decide which runtime to use for loading via a `model_type` setting in the explainer `model-setting.json`. 

Also, worth noting that if the model is saved via `joblib`, we don't even need to know what library (`xgboost`, `lightgbm` etc) the model belongs to. We can just `joblib.load()` and use the model directly...

Reply:
Notes from offline discussion:
- Having a separate `model-settings.json` for the model e.g.
   ```
   explainer/
      explainer_artefact/
      model-settings.json
      model/
         model_artefact.bst
         model-settings.json
   ```
   
   would be problematic, since mlserver would recursively search `explainer/` and deploy the model itself too, due to the presence of the extra `model-settings.json`.

- With above in mind, decision to stick with the current `try/except` strategy for now. However, will remove the `lgb.Booster` block since this doesn't return an sklearn api compatible model (user should use `joblib` for `lightgbm` models).

- Open issue wrt discrepancy between `mlserver_xgboost` and `mlserver_lightgbm`.
- Open issue wrt improving loading of white-box models in `SKLearnRuntime`.